File: new/posterior.R
library(Bessel)

source('~/Desktop/FYP/Codes/new/prior.R')

posterior_latent_z<-function(x,yt,yu,delta,z,v,vt,vu,w,set){
    val1 = prior_svvg(x)
    val2 = likelihood(yt,yu,x,delta,z,v,vt,vu)
    val3 = variance_gamma(z,x,delta)
    return(val1+val2+val3)
}

posterior_latent_w<-function(x,yt,yu,delta,z,v,vt,vu,w,set){
    val1 = prior_svvg(x)
    val2 = aux_g(v,w,x,delta,vt,vu,set)
    return(val1+val2)
}
posterior_latent_vt<-function(x,yt,yu,delta,z,v,vt,vu,w,set){
    val1 = sum(post_vt(vt,vu,x,delta))
    val2 = sum(aux_g(v,w,x,delta,vt,vu,set))
    val3 = sum(likelihood(yt,yu,x,delta,z,v,vt,vu))
    return(val1+val2+val3)
}

posterior_latent_vs<-function(x,yt,yu,delta,z,v,vt,vu,w,set){
    val1 = aux_g(v,w,x,delta,vt,vu,set)
    val2 = likelihood(yt,yu,x,delta,z,v,vt,vu)
    return(val1+val2)
}



posterior<-function(x,yt,yu,delta,z,v,vt,vu,w,set){

    val1 = prior_svvg(x)  # log form
    #print(val1)
    val2 = sum(variance_gamma(z,x,delta))    # log form
    #print(val2)
    val3 = sum(post_vt(vt,vu,x,delta))   # log form
    #print(val3)
    val4 = sum(aux_g(v,w,x,delta,vt,vu,set))
    #print(val4)
    val5 = sum(likelihood(yt,yu,x,delta,z,v,vt,vu))
    #print(val5)
    return(val1+val2+val3+val4+val5)
}

likelihood<-function(yt,yu,x,delta,z,v,vt,vu){
    v_bar = (vt-vu-x$k*x$v*delta+x$k*v)/x$sigma_v
    mn = yu+x$mu*delta+z+x$rho*v_bar
    dev = (1-x$rho^2)*v

    val = dnorm(yt,mean = mn,sd = sqrt(dev),log=T)
    
    return(val)
}

variance_gamma<-function(z,x,delta){
	a = delta
	sigma = x$sigma_j*sqrt(delta)
	val1 = 2*exp(x$gamma*z^2/sigma)/(a^(delta/a)*sqrt(2*pi)*gamma(delta/a))
	val2 = (z^2/(x$gamma^2+2*sigma^2/a))^(delta/(2*a)-0.5)
	v = z*sqrt(x$gamma^2+2*sigma^2/a)/sigma^2
	nu = delta/a-0.5
	val3 = BesselK(v,nu)
	return(log(val1*val2*val3)) 

}

post_vt<-function(vt,vu,x,delta){
	d = 4*x$k*x$v/x$sigma_v^2
	val1 = x$sigma_v^2*(1-exp(-x$k*delta))/(4*x$k)
	tau = 4*x$k*exp(x$k*delta)*vu/(x$sigma_v^2*(1-exp(-x$k*delta)))
	return(dchisq(vt,df=d,ncp = tau,log=T)+log(val1))
}

aux_g<-function(v,w,x,delta,vt,vu,set){
    #print(v[1:10])
    n = length(v)
    I_c = rep(1,n)
    I = rep(0,n)
    for(i in 1:n)
        if(v[i]<set$max_v)
            if(v[i]>set$min_v)
                if(w[i]<set$max_w)
                    if(w[i]>set$min_w){
                        I[i]= 1
                        I_c[i]= 0
                    }
    #print('hi0')
    temp = phi(w,x,delta,vt,vu)
    #print('hi')
    g = cos(v*w)*Re(temp)+sin(v*w)*Im(temp)
    #print('hi1')
    p = abs(g)*I+abs(g)*exp(-v-w)*I_c
    #print('hi3')
    #print(p)
    return(log(p))
}

phi<-function(y,x,delta,vt,vu){
    d = 4*x$k*x$v/x$sigma_v^2
    nu = sqrt(x$k^2-2*x$sigma_v^2*y*complex(imaginary=1))
    val1 = nu*exp(-0.5*(nu-x$k)*delta)*(1-exp(-x$k*(delta)))/(x$k*(1-exp(-nu*delta)))
    val2 = exp((vu+vt)/x$sigma_v^2*(x$k*(1+exp(-x$k*delta))/(1-exp(-x$k*delta))-nu*(1+exp(-nu*delta))/(1-exp(-nu*delta))))
    v1 = sqrt(vt*vu)*4*nu*exp(-0.5*nu*delta)/(x$sigma_v^2*(1-exp(-nu*delta)))
    v2 = sqrt(vt*vu)*4*x$k*exp(-0.5*x$k*delta)/(x$sigma_v^2*(1-exp(-x$k*delta)))
    #print('start')
    #print(v1[10])
    #print(d)
    val3_1 = BesselI(v1,nu=0.5*d-1)
    #print(val3_1[2])
    val3_2 = BesselI(v2,nu=0.5*d-1)
    #print(val3_2[2])
    val3 = val3_1/val3_2
    #print(val3[2])
    return(val1*val2*val3)
}

File: new/c++/posterior.cpp
#include "include.h"
#include <vector>
#include <iostream>
#include <math.h>
#include <boost/math/special_functions/bessel.hpp>
#include <complex>
#include <complex_bessel.h>
#include <boost/math/distributions/non_central_chi_squared.hpp>

using namespace std;

double post_no_prior(double z,double vt,double vu,double v_star,double yt,double yu,double w){
    double val2 = likelihood(z,vt,vu,v_star,yt,yu);
    double val3 = variance_gamma(z);
    double val4 = aux_g(v_star,w,vt,vu);
    double val5 = transition(vt,vu);
    return val2+val3+val4+val5;
}
double posterior(){
    flag = 1;
    double val1 = prior();
    double sum= 0 ;
    for(int i =1;i<total;i++){
        sum += post_no_prior(z[i],v[i],v[i-1],v_star[i],price[i],price[i-1],w[i]);
        if(!flag){
            break;
        }
    }
    return sum+val1;
}

double prior(){
    double prior_mu,prior_gam,prior_sigj,prior_rho,prior_com;
    prior_mu = -pow(mu,2)/2;
    prior_gam = -gam;
    double rhob = (rho+1)/2;
    prior_rho = (1/7-1)*log(rhob);
    prior_sigj = -sigma_j;
    double d = 4.0*k*v_p/pow(sigma_v,2);
    if(0.5*d>dmax || 0.5*d<1)
        flag = 0;
    prior_com = -sigma_v-v_p;
    return(prior_mu+prior_gam+prior_rho+prior_sigj+prior_com);
}

double likelihood(double z,double vt,double vu,double v_star,double yt,double yu){
    double var = (1-pow(rho,2))*v_star;
    double v_bar = (vt-vu-k*v_p*delta+k*v_star)/sigma_v;
    double mn = yu + mu*delta + z + rho*v_bar;
    double val = -log(pow(var,0.5))-pow(yt-mn,2)/(2*var);
    return val;
}

double variance_gamma(double z){
    double alpha = delta*1.0;
    double sigma = sigma_j*sqrt(delta);
    double val1 = gam*pow(z,2)/sigma-log(alpha)*delta/alpha-log(tgamma(delta/alpha));
    double val2 = (delta/(2*alpha)-0.5)*log(pow(z,2)/(pow(gam,2)+2*pow(sigma,2)/alpha));
    double va = sqrt(pow(z,2)*(pow(gam,2)+2*pow(sigma,2)/alpha))/pow(sigma,2);
    complex<double> a(va,0);
    complex<double> temp;
    temp = sp_bessel::besselK(delta/alpha-0.5,a);
    if(norm(temp)==0||isnan(va)||isinf(va))
        flag=0;
    double val3 = log(temp.real());
    return val1+val2+val3;
}

double aux_g(double v,double w,double vt,double vu){
    int I = 0;
    if(v>min_v)
        if(v<max_v)
            if(w>min_w)
                if(w<max_w)
                    I=1;
    std::complex<double> val = phi(v,w,vt,vu);
    if(norm(val) == 0)
        flag=0;
    double g = cos(v*w)*val.real()+sin(v*w)*val.imag();
    double aux = abs(g)*I+abs(g)*exp(-v-w)*(1-I);
    
    return(log(aux));
}

std::complex<double> phi(double v,double w,double vt,double vu){
    double de = 1.0*delta;
    double com = -2*pow(sigma_v,2)*w;
    complex<double> nu(pow(k,2),com);
    nu = sqrt(nu);
    double d = 0.5*4*k*v/pow(sigma_v,2)-1;
    complex<double> v1 = sqrt(vt*vu)*4.0*nu*exp(-0.5*nu*de)/(pow(sigma_v,2)*(1.0-exp(-nu*de)));
    complex<double> v2 = sqrt(vt*vu)*4.0*k*exp(-0.5*k*de)/(pow(sigma_v,2)*(1.0-exp(-k*de)));
    complex<double> den = sp_bessel::besselI(d,v2);
    complex<double> val1(0,0);
    if(norm(den)==0||isinf(d)||isnan(d)||isinf(v1.real())||isinf(v1.imag())||isinf(v2.real())||isinf(v2.imag())){
        flag = 0;
        return val1;
    }
    else{
        val1 = sp_bessel::besselI(d,v1)/den;
        val1 = val1*nu*exp(-0.5*(nu-k)*de)*(1.0-exp(-k*de))/(k*(1.0-exp(-nu*de)));
        val1 = val1*exp((vu+vt)/pow(sigma_v,2)*(k*(1.0+exp(-k*de))/(1.0-exp(-k*de))-nu*(1.0+exp(-nu*de))/(1.0-exp(-nu*de))));
    }
    return(val1);
}

double transition(double vt,double vu){
    double ncp = 4.0*k*exp(-k*delta)*vu/(pow(sigma_v,2)*(1-exp(-k*delta)));
    double  d = 4*k*k/pow(sigma_v,2);

    double val;
    
    if(isinf(vt)||isinf(vu)||isnan(vt)||isnan(ncp)){
        val = 0;
        flag = 0;
    }
    else{
        boost::math::non_central_chi_squared_distribution<long double> myNonCentralChiSquared(d, ncp);
        val = pdf(myNonCentralChiSquared,vt);
    }
    return log(val*pow(sigma_v,2)*(1-exp(-k*delta))/(4*k));

}




